kedro/io/data_catalog.py

```diff
@@ -615,15 +616,50 @@ def add_transformer(
                 )
             self._transformers[data_set_name].append(transformer)
 
-    def list(self) -> List[str]:
-        """List of ``DataSet`` names registered in the catalog.
+    def list(self, regex_search: Optional[str] = None) -> List[str]:
+        """
+        List of ``DataSet`` names registered in the catalog if called with
+        default arguments, this can be filtered by providing an optional regular
+        expression which will only return matching keys.
 
+        Args:
+            regex_search: An optional regular expression which can be provided
+                to limit the data sets returned by a particular pattern.
         Returns:
-            A List of ``DataSet`` names, corresponding to the entries that are
-            registered in the current catalog object.
+            A list of ``DataSet`` names available which match the
+            `regex_search` criteria if provided. All data sets are returned
+            by default.
+
+        Raises:
+            SyntaxError: When an invalid regex filter is provided.
+
+        Example:
+        ::
 
+            >>> io = DataCatalog()
+            >>> # get data sets where the substring 'raw' is present
+            >>> raw_data = io.list(regex_search='raw')
+            >>> # get data sets which start with 'prm' or 'feat'
+            >>> feat_eng_data = io.list(regex_search='^(prm|feat)')
+            >>> # get data sets which end with 'time_series'
+            >>> models = io.list(regex_search='.+time_series$')
         """
-        return list(self._data_sets.keys())
+
+        if regex_search is not None:
+            try:
+                pattern = re.compile(regex_search, flags=re.IGNORECASE)
+            except re.error:
+                raise SyntaxError(
+                    f"Invalid pattern regular expression provided: `{regex_search}`"
+                )
+            working_data_sets = {
+                k: v for k, v in self._data_sets.items() if pattern.search(k)
+            }
+
+        else:
+            working_data_sets = self._data_sets
+
+        return list(working_data_sets.keys())
```

Comment:
We can simplify this now that `get_object` is descoped, and I believe a regex of empty string would match everything.
```suggestion

        if not regex_search:
            return list(self._data_sets.keys())

        try:
            pattern = re.compile(regex_search, flags=re.IGNORECASE)
        except re.error:
            raise SyntaxError(
                f"Invalid pattern regular expression provided: `{regex_search}`"
            )

        matching_data_sets = {
            k: v for k, v in self._data_sets.items() if pattern.search(k)
        }
        return list(matching_data_sets.keys())
```

Reply:
In fact, you no longer need to build a dictionary in the `regex` case, something like:

```python
matching_data_sets [name for name in self._data_sets if pattern.search(name)]
```

should suffice 🥳

Reply:
I've taken bits of both your suggestions. @lorenabalan good shout on the empty string edge case, uncaught it will match all datasets. I've caught this and now log a warning too.